Container-based proto compilation

n0shut opened the pull request:
Hi @robshakir @marcushines 

I would like to propose a hermetic way to compile protos used in this repository which is based on the container approach.

Currently, the compile-protos.sh script has no pinned dependencies resulting in users defaulting to whatever latest/oldest version they have installed on their machines when updating the protos.

Not only it adds unnecessary churn to the PRs, it also makes the process overly complicated. The dev machine should have protos installed in the specified directories and we see that people struggle with that - #134.

In the newly added compile-protos-docker.sh I introduced a container-based workflow where I leverage ghcr.io/srl-labs/protoc:0.0.3 that has pinned dependencies for protoc and the relevant grpc plugins - https://github.com/srl-labs/protoc-container/blob/main/Dockerfile

Using this container every user of gnmi repo will be guaranteed to compile the protos in one and only possible way, without the need to keep local dependencies and maintain their lifecycle.

At the time of this writing the latest version of protoc and plugins were used. I hope you will find that useful.

robshakir commented on compile-protos-docker.sh:
```diff
@@ -0,0 +1,61 @@
+
+DOCKER_CMD="docker run -v $(pwd):/in \
+  -v $(pwd)/proto:/in/github.com/openconfig/gnmi/proto \
+  ghcr.io/srl-labs/protoc:0.0.3"
```
For my edification, how does this map to the `protoc` version that is in the container?

hellt replied:
@robshakir it doesn't direclty map to the protoc version, as there are more than just `protoc` in the package (namely grpc plugins for both py and go version).
Therefore, the idea is that `ghcr.io/srl-labs/protoc:0.0.3` is tagged in a semver fashion and the image tag matches the tag in the git repo for which the Dockerfile reveals the used versions: https://github.com/srl-labs/protoc-container/blob/v0.0.3/Dockerfile

robshakir replied:
ACK - no problem. I was just wondering whether it is possible to directly map these together so that it's clearer, but it's written into the `pb.go` as to what versions are used so I don't see this is a huge issue.

The only downside here is that we stay pinned to that `0.0.3` version until an explicit change is made. If there's a semver convention here - could we consider tagging something like `1.x.x-latest`? (It'd be nice to stay close to HEAD for the generated files.)

n0shut replied:
@robshakir good call, I see that the upstream changed the logic to include protoc__protoc-gen-go version in the image tag, so that it clear which sw components versions are used.

https://github.com/srl-labs/protoc-container#versioning

I have therefore update the branch in [4a11050](https://github.com/openconfig/gnmi/pull/136/commits/4a11050420ed9f7bd5a7f970134f2a2ccadfb855) to reflect that.

Now it is explicit which version of protoc and protoc-gen-go are used in the image.

n0shut replied:
I can also port it to gnoi repo where it is even more of a pressing issue.

robshakir reviewed (commented):
This looks great, thanks @n0shut. One quick question in the review.

n0shut commented on compile-protos-docker.sh:
```diff
@@ -0,0 +1,65 @@
+## Generating Go bindings ##
+#
+
+# generating testing/fake/proto/fake.proto
+${DOCKER_CMD} \
+  bash -c "${PROTOC_GO_CMD} testing/fake/proto/fake.proto"
+
+# generating proto/gnmi/gnmi.proto
+${DOCKER_CMD} \
+  bash -c "${PROTOC_GO_CMD} proto/gnmi/gnmi.proto"
+
+# generating proto/collector/collector.proto
+${DOCKER_CMD} \
+  bash -c "${PROTOC_GO_CMD} proto/collector/collector.proto"
+
+# generating proto/gnmi_ext/gnmi_ext.proto
+${DOCKER_CMD} \
+  bash -c "${PROTOC_GO_CMD_NOGRPC} proto/gnmi_ext/gnmi_ext.proto"
+
+# generating proto/target/target.proto
+${DOCKER_CMD} \
+  bash -c "${PROTOC_GO_CMD_NOGRPC} proto/target/target.proto"
```
One thing I am not particularly happy with is that this script generated bindings for all the listed protos, which might be not desirable I think? For example, if only gnmi_ext has changed and you want to regenerate its bindings it would be cool to have a command that does just that.

What I normally do myself is writing a Make file with separate targets, something like this:

* `make gen-all` - generate bindings for all listed protos
* `make gen-fake.proto` - generate bindings for fake.proto
* etc

But since I do not see a Makefile in this repo I didn't want to introduce right away.

If make is not to ones liking, it can be a bash with functions, simulating the same behavior.

n0shut commented:
Hi @robshakir 
I have updated this PR by introducing `./run.sh` script that embeds purpose-built functions for each lang/proto.
I have updated the initial PR comment to reflect that, but in essense what you can now do is:

```bash
# to compile all protos
./run.sh compile-all

# to compile selected protos
./run.sh compile-go-gnmi-ext
```

hellt commented:
tsk tsk tsk
still using this user unfriendly compile_protos.sh that has no reproducibility and heavily dependent on hosts' configuration

robshakir commented:
Thanks for this -- let me look again this week. Unfortunately, I was on an extended leave during the period you previously requested review.

hellt commented:
No worries @robshakir, I didn't mean to push, just a slight nudge. I can update it with the updated versions of protoc and go plugins

robshakir commented on run.sh:
```diff
@@ -0,0 +1,122 @@
+set -o pipefail
+
+# container image tag, to see all available tags go to
+# https://github.com/srl-labs/protoc-container/pkgs/container/protoc
```
Is this container something that you expect to continue to be updated?

n0shut replied:
yes, I just bumped it to a more recent version tuple [227be07](https://github.com/openconfig/gnmi/pull/136/commits/227be073d1d6105d6e67548e74317bd7617a39a8)

robshakir commented on run.sh:
```diff
@@ -0,0 +1,122 @@
+function compile-go-gnmi {
+    ${DOCKER_CMD} ash -c "${PROTOC_GO_CMD} proto/gnmi/gnmi.proto"
+    echo "finished Go compilation for gnmi.proto"
+    
```
(nit) is this newline needed (here and below)?

n0shut replied:
fixed, thanks

robshakir reviewed (commented):
I'd propose to also leave `compile_protos.sh` present as well -- since there are some environments where security restrictions mean that docker cannot be run (e.g., locations that do not allow a virtual machine to be run).

robshakir commented:
@marcushines - PTAL too.

n0shut commented:
compile_protos.sh brought back, @robshakir

n0shut commented:
closing as stale
